schema_editor/test/spec/schemas/schemas-service.spec.js

```diff
@@ -148,4 +176,69 @@ describe('ase.schemas:Schemas', function () {
             message: jasmine.stringMatching('Diff')
         }]));
     });
+
+    // NOTE: This test must be updated when new field types are added.
+    it('should preserve all information when serializing and deserializing schemas', function () {
+        // In practice this means that the serialize and deserialize methods are inverses
+        var f = Schemas.definitionFromSchemaFormData;
+        var b = Schemas.schemaFormDataFromDefinition;
+
+        var schemaFormData = [{
+            fieldType: 'selectlist',
+            isSearchable: false,
+            displayType: 'checkbox',
+            isRequired: false,
+            fieldTitle: 'Select',
+            fieldOptions: ['opt1', 'opt2'],
+            propertyOrder: 0 // The serializer uses the ordering of the incoming form data to
+                             // determine and set the propertyOrder value, since this is how
+                             // JSON-Editor expresses the value. Therefore this value must match
+                             // the ordering of the array in order to ensure that the output
+                             // matches the input, otherwise the serializer will assume that the
+                             // ordering has changed and update the propertyOrder field, causing
+                             // the test to fail.
+        },{
+            fieldType: 'image',
+            isSearchable: false,
+            isRequired: true,
+            fieldTitle: 'Image',
+            propertyOrder: 1
+        },{
+            fieldType: 'text',
+            isSearchable: false,
+            isRequired: false,
+            fieldTitle: 'Text',
+            textOptions: 'datetime',
+            propertyOrder: 2
+        }];
+
+        // Transform back and forth a few times
+        var output = b(f(b(f(b(f(b(f(b(f(b(f(schemaFormData))))))))))));
```

Comment:
Yeah...I might have gotten a bit carried away with the nesting, but at least I'm _very_ confident they are inverses!